### PR TITLE
add noninitial plastic strain property

### PR DESCRIPTION
**Pull request by naliboff**
This pull request enables the visco plastic strain invariant particle property to track noninitial plastic strain, and also (hopefully) improves the code structure and documentation. I tested this locally with a new test case, but we will see what happens with the existing tests.

* [X] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).
* [X] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: source/particle/property/viscoplastic_strain_invariants.cc [hunk 1]
```diff
@@ -123,39 +129,46 @@ namespace aspect
...
+         * (5) If only viscous strain is tracked, it will be in the first data position.
+         * (6) If both viscous and plastic strain are tracked, viscous strain will be in the second data position
+         * If these assumptions change in the future, they will need to be updated.
          * */
+
         if (this->introspection().compositional_name_exists("plastic_strain") && plastic_yielding == true)
-          data[data_position] = new_strain;
+          data[data_position] += strain_update;
 
         if (this->introspection().compositional_name_exists("viscous_strain") && plastic_yielding == false)
```
Comment (anne-glerum) [inline source/particle/property/viscoplastic_strain_invariants.cc:155]: Thank you for adding the list of assumptions above, it's much clearer! Could you please also add for each of the below conditions a comment that explains what situation we're in? I'll add some examples below. I think that will make it easier to change the code if we ever change the assumptions.

File: source/particle/property/viscoplastic_strain_invariants.cc [hunk 2]
```diff
@@ -123,39 +129,46 @@ namespace aspect
...
-          data[data_position+(n_components-1)] = new_strain;
+          {
+            if (n_components == 1)
```
Comment (anne-glerum) [inline source/particle/property/viscoplastic_strain_invariants.cc:157]: `// Not yielding and only one field, which tracks the viscous strain`

File: source/particle/property/viscoplastic_strain_invariants.cc [hunk 3]
```diff
@@ -123,39 +129,46 @@ namespace aspect
...
+              data[data_position] += strain_update;
+
+            if (n_components == 2)
```
Comment (anne-glerum) [inline source/particle/property/viscoplastic_strain_invariants.cc:160]: `// Not yielding, two fields, second is viscous strain.`

File: source/particle/property/viscoplastic_strain_invariants.cc [hunk 4]
```diff
@@ -123,39 +129,46 @@ namespace aspect
...
+              data[data_position+(n_components-1)] += strain_update;
+
+            if (n_components == 3)
```
Comment (anne-glerum) [inline source/particle/property/viscoplastic_strain_invariants.cc:163]: `// Not yielding, three fields (plastic strain, viscous strain, nonintial plastic strain).`

**Review by anne-glerum — CHANGES_REQUESTED** [review]
Thank you John, that's a lot clearer! Can you add a test that tracks the noninitial strain on particles?

**Comment by naliboff** [comment]
@anne-glerum - Thanks for the review and I've hopefully addressed all of the comments. Good idea to add a description above each scenario, which will make this much easier to update later on. Please let me know if any additional changes are needed.

File: tests/visco_plastic_yield_noninitial-plastic-strain_particles.prm [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# A test to check whether noninitial plastic strain is correctly tracked
+# on particles, which are also used to track plastic and viscous strain.
+# The boundary conditions (see tests/visco_plastic_yield_plastic_viscous_strain_weakening.prm)
+# produce a incremental strain of 1e-5 per time step and material that is always yielding.
```
Comment (anne-glerum) [inline tests/visco_plastic_yield_noninitial-plastic-strain_particles.prm:4]:
```suggestion
# produce an incremental strain of 1e-5 per time step and material that is always yielding.
```

**Review by anne-glerum — CHANGES_REQUESTED** [review]
Thanks for adding the test! Could you please add a line about the noninitial_plastic_strain in the description of the particle property at the bottom of the cc file? Sorry I didn't see that before.

File: tests/visco_plastic_yield_noninitial-plastic-strain_particles.prm [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+# viscous strain: 0. As a result, the viscous strain should always have a value of 0 throughout
+# the model run. Respectively, after 3 time steps the plastic and noninitial plastic values
+# should be 4e-5 and 3e-5.
+include $ASPECT_SOURCE_DIR/tests/visco_plastic_yield_plastic_viscous_strain_weakening.prm
+
+set End time           = 3000
+set Maximum time step  = 1000
+set Output directory   = visco_plastic_yield_noninitial-plastic-strain_particles
+
+# Particles mapped to the Compositional fields used to track plastic, noninitial plastic, and viscous strain.
```
[the remark this answers is not in the document]
Comment (naliboff) [reply tests/visco_plastic_yield_noninitial-plastic-strain_particles.prm:15]: Yep, it took me a bit of looking to figure out this out as well!

**Comment by naliboff** [comment]
@anne-glerum - Thanks for the additional review and suggestions! I've incorporated all of the suggestions and also simplified the logic in the viscous strain update section, per your recommendation. Assuming the tests still pass, let me know if you would like any additional changes.

**Review by anne-glerum — APPROVED** [review]
Great thank you!

**Comment by anne-glerum** [comment]
Do you think this warrants a change log entry?

**Comment by naliboff** [comment]
@anne-glerum - Good point, I added a change log entry.

**Review by anne-glerum — APPROVED** [review]

**Comment by anne-glerum** [comment]
Thanks John!
